wence/complex

wence- opened the pull request:
This is the result of the complex sprint, with commits cleaned up. Firedrake master passes with this branch (https://jenkins.ese.ic.ac.uk:1080/blue/organizations/jenkins/firedrake/detail/wence%2Ftest-complex-tsfc/2/pipeline)

It needs one mini-patch in Firedrake (to set the scalar_type in tsfc parameters correctly).

I think this is ready for review.

wence- commented on gem/impero_utils.py:
```diff
@@ -316,3 +316,84 @@ def recurse_top(o):
+
+
+def dtype_sort_key(dtype):
+    return 1 if dtype.kind == 'c' else -1
```
Should use `numpy.find_common_dtype`

wence- commented on gem/refactorise.py:
```diff
@@ -169,7 +169,7 @@ def stop_at(expr):
     for expr in compounds:
         summands = traverse_sum(expr, stop_at=stop_at)
-        if len(summands) <= 1 and not isinstance(expr, Conditional):
+        if len(summands) <= 1 and not isinstance(expr, (Conditional, MathFunction)):
```
MathFunction has a special handler (to deal with conj/real/imag).

wence- commented on gem/refactorise.py:
```diff
@@ -211,6 +211,23 @@ def stop_at(expr):
+        result.ordering = child_ms.ordering.copy()
+        return result
+    else:
+        return _collect_monomials.dispatch(MathFunction.mro()[1])(expression, self)
```
singledispatch version of `super()`.

wence- commented on tsfc/driver.py:
```diff
@@ -50,14 +50,7 @@ def compile_form(form, prefix="form", parameters=None, interface=None, coffee=Tr
-        logger.warning("Disabling whole expression optimisations"
-                       " in GEM for supporting complex mode.")
-        parameters = parameters.copy()
-        parameters["mode"] = 'vanilla'
```
Refactoriser now handles complex nodes.

wence- commented on tsfc/loopy.py:
```diff
@@ -81,13 +110,13 @@ def generate(impero_c, args, precision, scalar_type, kernel_name="loopy_kernel",
     # Create arguments
     data = list(args)
-    for i, temp in enumerate(impero_c.temporaries):
+    for i, (temp, dtype) in enumerate(assign_dtypes(impero_c.temporaries, scalar_type)):
```
Type propagation.

wence- commented on tsfc/loopy.py:
```diff
@@ -241,11 +270,6 @@ def _expression_power(expr, ctx):
-    from tsfc.coffee import math_table
-
-    math_table = math_table.copy()
-    math_table['abs'] = ('abs', 'cabs')
```
Loopy mode now has a separate math table.

wence- commented on tsfc/loopy.py:
```diff
@@ -329,10 +354,10 @@ def _expression_conditional(expr, ctx):
-    r = round(v, 1)
-    if r and abs(v - r) < parameters.epsilon:
+    r = numpy.round(v, 1)
+    if r and numpy.abs(v - r) < parameters.epsilon:
```
`math.isnan/round` don't work with numpy types.

wence- commented on tsfc/parameters.py:
```diff
@@ -14,8 +14,11 @@
+    "scalar_type": numpy.dtype(numpy.float64),
+
+    # So that tests pass (needs to match scalar_type)
+    "scalar_type_c": "double",
```
This is a really dumb way to do this (scalar_type_c can be expunged when coffee backend goes).

wence- commented on gem/impero_utils.py:
```diff
@@ -316,3 +317,80 @@ def recurse_top(o):
+@_assign_dtype.register(gem.LogicalAnd)
+@_assign_dtype.register(gem.LogicalOr)
+def _assign_dtype_logical(expression, self):
+    return numpy.bool
```
Not sure these will ever cons temporaries, but we may as well handle them.

wence- replied:
turns out they do!

wence- commented on gem/impero_utils.py:
```diff
@@ -316,3 +317,80 @@ def recurse_top(o):
+    if expression.name in {"real", "imag"}:
+        return self.real_type
+    else:
+        return maxtype(map(self, expression.children))
```
Is this the safe thing to do? I don't think we support any mathfunctions that complexify real values in complex mode right now.

miklos1 replied:
What about square root of negatives?

wence- replied:
Oh yeah. Probably what I should do is handle the entire "math-table" one by one (which suggests migrating this mapper into the loopy codegen part [this is OK, since I didn't fix up the coffee backend]).

wence- replied:
Done.

wence- commented:
And now with passing firedrake again (real mode): https://jenkins.ese.ic.ac.uk:1080/blue/organizations/jenkins/firedrake/detail/wence%2Ftest-complex-tsfc/4/pipeline/67

wence- commented:
Reminder before merging to fix the `simplify_abs` visitor:
```python
@_simplify_abs.register(Sqrt)
def _simplify_abs_sqrt(o, self, in_abs):
    # Square root is always non-negative
    return ufl_reuse_if_untouched(o, self(o.ufl_operands[0], False))
```

miklos1 commented on gem/impero_utils.py:
```diff
@@ -316,3 +317,80 @@ def recurse_top(o):
+@_assign_dtype.register(gem.LogicalAnd)
+@_assign_dtype.register(gem.LogicalOr)
+def _assign_dtype_logical(expression, self):
+    return numpy.int8
```
Why not `numpy.bool_`?

wence- replied:
The dtype mapping in loopy generates `bool` for `numpy.bool_` which is not a C type.

wence- replied:
I fixed this in inducer/compyte#30, but it is not yet backmerged into the firedrake loopy branch.

miklos1 commented on gem/impero_utils.py:
```diff
@@ -316,3 +317,80 @@ def recurse_top(o):
+        mapper.real_type = scalar_type.type(0).real.dtype
+    else:
+        mapper.real_type = scalar_type
+    return [(e, mapper(e)) for e in expressions]
```
This only needs GEM and numpy, but not the Impero AST.  Nevertheless, since it's so closely coupled to code generation, I think it's fine that this visitor stays here.

wence- replied:
I moved it close to the mathtable in loopy.py now.

miklos1 commented on gem/refactorise.py:
```diff
@@ -211,6 +211,23 @@ def stop_at(expr):
+        child_ms, = map(self, expression.children)
+        result = MonomialSum()
+        for k, v in child_ms.monomials.items():
+            result.monomials[k] = MathFunction(name, v)
```
So you only propagate these complex functions to the "rest" factors, which means you assume that "atomics" are real.  This is true for argument factorisation (which is the main use of the refactoriser), but I'm not sure that this would work correctly with tensor mode (which uses the refactoriser for a different refactorisation).

I think this special assumption should at least be noted in a comment.

wence- replied:
`atomics` in this case (arguments) are the basis tabulations, yes? I guess the "right thing" would be to check what atomics are and propagate to both (if necessary). Or one could have a flag in the refactoriser that says "are atomics real?". WDYT?

wence- replied:
Added a comment.

miklos1 commented:
Lawrence Mitchell <notifications@github.com> writes:
> @wence- commented on this pull request.
> +@_collect_monomials.register(MathFunction)
> +def _collect_monomials_mathfunction(expression, self):
> +    name = expression.name
> +    if name in {"conj", "real", "imag"}:
> +        # These are allowed to be applied to arguments, and hence must
> +        # be dealt with specially. Just push the function onto each
> +        # entry in the monomialsum of the child.
> +        child_ms, = map(self, expression.children)
> +        result = MonomialSum()
> +        for k, v in child_ms.monomials.items():
> +            result.monomials[k] = MathFunction(name, v)
>
> `atomics` in this case (arguments) are the basis tabulations, yes? I
> guess the "right thing" would be to check what atomics are and
> propagate to both (if necessary). Or one could have a flag in the
> refactoriser that says "are atomics real?". WDYT?

"Atomics" are defined by the classifier function, and the classifier
functions are defined by the optimisation modes.

I just looked up the source code, and the real assumption is actually
fine for all our optimisation modes.  Vanilla does not call
refactoriser, tensor defines atomics as tabulation matrices with a
quadrature index, coffee and spectral modes define atomics as tabulation
tensors with argument indices or delta nodes with argument indices.

I guess adding a comment about the real assumption suffices for now.

wence- commented:
> Reminder before merging to fix the simplify_abs visitor:

Now done.

miklos1 commented on tsfc/loopy.py:
```diff
@@ -23,6 +22,116 @@
+    'atan2': ('atan2', None),
+    'min_value': ('min', None),
+    'max_value': ('max', None)
+}
```
This table now feels almost unnecessary.

- The real and complex function names are always the same for those functions which have both versions.
- ... and this target function name is in most cases the same as the source function name, except for `power`/`pow`, `ln`/`log`, `atan_2`/`atan2`, `min_value`/`min`, and `max_value`/`max`.
- Why are there two source names for atan2 by the way?  That looks like an inconsistency somewhere else in the Firedrake stack.
- Power, min value, and max value have their own GEM nodes anyway, so they need not make use of this table.
- That leaves `ln`/`log` alone where a name translation is necessary, plus the dubious atan2 case.

wence- replied:
I checked `atan_2` is just a mistake. I refactored to remove this table. That also fixes a bug where one couldn't use `erf` (or similar) in complex mode even if the argument was wrapped in `real` to ensure realness.

wence- commented:
I think I've now addressed all the comments. I made enough changes that I need to run through with the real-mode version of firedrake to check, but I hope that this now at least passes tsfc tests. With the appropriate loopy branch it passes tsfc tests when the scalar type is complex too.

miklos1 commented on tsfc/coffee.py:
```diff
@@ -213,7 +213,7 @@ def _expression_division(expr, parameters):
 math_table = {
     'sqrt': ('sqrt', 'csqrt'),
-    'abs': ('fabs', 'cabs'),
+    'abs': ('fabs', 'abs'),
```
COFFEE mode probably not very relevant anymore, but this change does not look right.

wence- replied:
I cherry-picked from the sprint branch (particularly ff5193e66408d) perhaps @sv2518 can comment on why this change was needed?

wence- replied:
Oh, I guess it's because the slate-mode kernels were being compiled in C++ mode, for which `abs` is correct. But we decided not to port slate until it is migrated to the loopy codegen path anyway, so this change is moot.

miklos1 replied:
Oh yes, complex is quite different in C and in C++...

wence- replied:
Dropped this part.

miklos1 commented:
So I think this is fine to be merged, though I note that the code seems to be growing in a direction where one should consider introducing GEM nodes `Real`, `Imag`, and `Conj`.  There are some 3 cases when one checks whether the name of a `MathFunction` is `in {'real', 'imag', 'conj'}`.  This change would probably increase number of code lines at this point, but it seems to be the better code organisation if this trend continues.

wence- commented:
The abs-simplification change was incomplete and I think I've now fixed it. I'm a bit unhappy that it involves pushing this `complex_mode` flag through so many places though.
